app/src/main/java/tttTDD/Java/Game.java

```diff
@@ -0,0 +1,48 @@
+package tttTDD.Java;
+
+import tttTDD.Java.Interfaces.Player;
+
+public class Game {
+    Board board;
+    Player player1;
+    Player player2;
+    Player currentPlayer;
+    boolean gameCompleted;
+    String gameWinner;
+
+    public Game(GameConfiguration gameConfig) {
+        board = new Board(gameConfig.getBoardSizeConfiguration());
+        initializePlayers(gameConfig.isComputerPlaying());
+        currentPlayer = player1;
+        gameCompleted = false;
+    }
+    // humanPlayer.move(int) => return int
+    // list all the things that it does and all the things that it needs
+    public boolean performMove(int playerInput) {
+        if (!currentPlayer.isComputer()) {
+            board.updateBoard(playerInput, currentPlayer.getMarker());
+        } else {
+            board.updateBoard(board.getFirstAvailableMove(), currentPlayer.getMarker());
```

Comment:
The [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) is usually discussed in the context of the interfaces we build for end users, but I think it's good to apply quite generally.

So, for example, if I were consuming this method (and hadn't authored it), I would expect, based on the function signature and name, that the move I pass in is the one that is made, or to get a `false` return value, which I would assume implied that no move was made at all.

I would be quite astonished to find that actually a completely arbitrary move would instead be made every other time that I call this.

This can be partially alleviated by documenting the behavior (and proper use) of the method in Javadoc, but if I'm in a hurry, I'm likely to make an (incorrect) assumption that this is more straightforward than it is.

Reply:
Great share. I think that message has largely been lost lately, even with all the talk around mitigating/removing side effects as a whole.